File: src/main/java/com/slavsquatsuperstar/mayonez/KeyInput.java
package com.slavsquatsuperstar.mayonez;

import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

public class KeyInput extends KeyAdapter {

	// Key Fields
	private boolean[] keys = new boolean[256];

	// KeyListener Methods

	@Override
	public void keyPressed(KeyEvent e) {
		int code = e.getKeyCode();
		keys[code] = true;
	}

	@Override
	public void keyReleased(KeyEvent e) {
		int code = e.getKeyCode();
		keys[code] = false;
	}

	// Getters and Setters

	public boolean keyDown(int keyCode) {
		return keys[keyCode];
	}

	/**
	 * Whether the specified {@link KeyMapping} is pressed.
	 * 
	 * @param keyName The name of the {@link KeyMapping}.
	 */
	public boolean keyDown(String keyName) {
		for (KeyMapping m : KeyMapping.values())
			if (m.toString().equalsIgnoreCase(keyName)) // if the desired mapping exists
				return keys[m.keyCode];
		return false;
	}

	public int getAxis(String axisName) {
		for (KeyAxis a : KeyAxis.values())
			if (a.toString().equalsIgnoreCase(axisName))
				return a.value();
		return 0;
	}

	/*
	 * Enum Definitions
	 */

	/**
	 * Stores two keys intended to perform opposite actions.
	 */
	enum KeyAxis {
		
		VERTICAL(KeyMapping.DOWN, KeyMapping.UP), HORIZONTAL(KeyMapping.RIGHT, KeyMapping.LEFT);

		private int posKey, negKey;

		private KeyAxis(KeyMapping posKey, KeyMapping negKey) {
			this.posKey = posKey.keyCode();
			this.negKey = negKey.keyCode();
		}

		/**
		 * @return 1 if the positive key is pressed.<br/>
		 *         -1 if the negative key is pressed.<br/>
		 *         0 if the both or neither key is pressed.<br/>
		 */
		public int value() { // TODO keyboard acceleration?
			// "vector "method to make sure keys don't override each other
			int negComp = (Game.keyboard().keyDown(negKey)) ? -1 : 0;
			int posComp = (Game.keyboard().keyDown(posKey)) ? 1 : 0;
			return negComp + posComp;
		}

		@Override
		public String toString() {
			return name().substring(0, 1).toUpperCase() + name().substring(1).toLowerCase();
		}
	}

	/**
	 * Associates a virtual key code to a name.
	 */
	enum KeyMapping {

		// TODO read from file to assign keybinds
		UP(KeyEvent.VK_W), DOWN(KeyEvent.VK_S), LEFT(KeyEvent.VK_A), RIGHT(KeyEvent.VK_D), EXIT(KeyEvent.VK_ESCAPE),
		SPACE(KeyEvent.VK_SPACE), SHIFT(KeyEvent.VK_SHIFT);

		private int keyCode; // TODO allow for multiple?
//		private int[] keyCodes;

		private KeyMapping(int keyCode) {
			this.keyCode = keyCode;
		}

//		private KeyMapping(int... keyCodes) {
//			this.keyCodes = keyCodes;
//		}

		public int keyCode() {
			return keyCode;
		}

//		public int[] getKeyCodes() {
//			return keyCodes;
//		}

		@Override
		public String toString() {
			return name().toLowerCase();
		}

	}

}

File: mayonez-base/src/main/kotlin/slavsquatsuperstar/mayonez/input/KeyInput.kt
package slavsquatsuperstar.mayonez.input

import org.lwjgl.glfw.GLFW.*
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent

/**
 * The receiver for all keyboard-related input events.
 *
 * @author SlavSquatSuperstar
 */
// TODO game loop / callback methods shouldn't be accessible from API (called statically)
object KeyInput : KeyAdapter() {

    // Key Fields
    private val keys = HashMap<Int, InputState>()

    // Game Loop Methods
    @JvmStatic
    fun endFrame() {
//        for (k in keys.keys)  {
//            if (keys[k] == InputState.PRESSED) keys[k] = InputState.HELD
//        }
    }

    /* Keyboard Callbacks */

    @JvmStatic
    @Suppress("unused")
    fun keyCallback(window: Long, key: Int, scancode: Int, action: Int, mods: Int) {
        when (action) {
            GLFW_PRESS -> setKey(key, InputState.PRESSED)
            GLFW_REPEAT -> setKey(key, InputState.HELD)
            GLFW_RELEASE -> setKey(key, InputState.RELEASED)
        }
    }

    override fun keyPressed(e: KeyEvent) { // Activates when ever a key is down
        println("pressed")
        val key = e.keyCode
        if (isKeyReleased(key)) setKey(key, InputState.PRESSED) // New key press -> set as pressed
        else setKey(key, InputState.HELD) // Continuous key press -> set as held
    }

    override fun keyReleased(e: KeyEvent) {
        setKey(e.keyCode, InputState.RELEASED) // Set as released
    }

    private fun setKey(keyCode: Int, state: InputState) {
        keys[keyCode] = state
    }

    /* Keyboard Getters */

    private fun isKeyReleased(keyCode: Int): Boolean = (keys[keyCode] == InputState.RELEASED || keys[keyCode] == null)

    private fun isKeyPressed(keyCode: Int): Boolean = keys[keyCode] == InputState.PRESSED

    private fun isKeyHeld(keyCode: Int): Boolean = keys[keyCode] == InputState.HELD

    @JvmStatic
    fun keyDown(keyCode: Int): Boolean = isKeyHeld(keyCode) || isKeyPressed(keyCode)

    @JvmStatic
    fun keyPressed(keyCode: Int): Boolean = isKeyPressed(keyCode) && !isKeyHeld(keyCode)

    /**
     * Returns whether any of the keys associated with the specified [KeyMapping] is continuously held down.
     *
     * @param keyName the name of the [KeyMapping]
     * @return if the specified key is down
     */
    @JvmStatic
    fun keyDown(keyName: String): Boolean {
        for (m in KeyMapping.values())
            if (m.name.equals(keyName, ignoreCase = true)) // if the desired mapping exists
                for (code in m.keyCodes) if (keyDown(code)) return true
        return false
    }

    /**
     * Returns whether any of the keys associated with the specified [KeyMapping] has been pressed.
     *
     * @param keyName the name of the [KeyMapping]
     * @return if the specified key was pressed
     */
    @JvmStatic
    fun keyPressed(keyName: String): Boolean {
        for (m in KeyMapping.values())
            if (m.name.equals(keyName, ignoreCase = true))
                for (code in m.keyCodes) if (keyPressed(code)) return true
        return false
    }

    @JvmStatic
    fun getAxis(axisName: String): Int {
        for (a in KeyAxis.values())
            if (a.toString().equals(axisName, ignoreCase = true))
                return a.value()
        return 0
    }

}